keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,295 @@
+
+## Summary
+
+Defaulting is a fundamental step in the processing of API objects in the request pipeline of the kube-apiserver. Defaulting happens during deserialization, i.e. after decoding of a versioned object, **but before** conversion to a hub type.
+
+Defaulting is implemented for most native Kubernetes API types and plays a crucial role for API compatibility when adding new fields. CustomResources do not support this natively. 
+
+Mutating admission webhooks can be used to partially replicate defaulting for incoming request payloads. But mutating admission webhooks do not run when reading from etcd.
+
+This KEP is about adding support for specifying default values for fields via OpenAPI v3 validation schemas in the CRD manifest. OpenAPI v3 has native support for a `default` field with arbitrary JSON values, for example: 
+
+```yaml
+properties:
+  foo:
+    type: string
+    default: "abc" 
+```
+
+This KEP proposes to apply these default values during deserialization, in the same way as native resources do. We assume _structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002).
+
+## Motivation
+
+* By far most native Golang based resources implement defaulting. CRD do not allow that, leading to unnatural, not Kubernetes-like APIs. This is bad for the ecosystem.
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequat as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002) makes defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults. Native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+  * undefined (`_, ok := obj[field]; !ok`)
+  * `nil` if the field not nullable.
```

Comment:
@liggitt this is maybe a key detail we have to think about. What about empty list/map?